File: plugins/delay/reverse-delay/reverse-delay.js
// Reverse Delay — record a window of length T, then play it back in
// reverse. As the write head sweeps forward and the read head sweeps
// backward through the same chunk, you hear the input rolled backwards
// — the trademark sound from Pink Floyd "Empty Spaces" and countless
// shoegaze records. Feedback re-injects the reversed wet into the line,
// piling up nested reversals into a swelling cloud.

function ReverseDelayProcessor() {
  // 4 s max @ 48 kHz.
  this.buf = new Array(192000);
  for (var k = 0; k < this.buf.length; k++) this.buf[k] = 0;
  this.bufLen = this.buf.length;
  this.writeIdx = 0;
  this.readOffset = 0;
}
ReverseDelayProcessor.parameterDescriptors = [
  { name: 'time',     label: 'Time (ms)', defaultValue: 600, minValue: 100, maxValue: 4000 },
  { name: 'feedback', label: 'Feedback',  defaultValue: 0.3, minValue: 0,   maxValue: 0.9 },
  { name: 'mix',      label: 'Mix',       defaultValue: 0.5, minValue: 0,   maxValue: 1 }
];
ReverseDelayProcessor.prototype.process = function(inputs, outputs, parameters) {
  var input = inputs[0][0], output = outputs[0][0];
  if (!input || !output) return true;
  var timeMs = parameters.time[0];
  var feedback = parameters.feedback[0];
  var mix = parameters.mix[0];
  var len = Math.max(64, Math.floor(sampleRate * timeMs / 1000));
  if (len > this.bufLen) len = this.bufLen;
  for (var i = 0; i < input.length; i++) {
    // Read N samples back, stepping backwards through the chunk.
    var idx = this.writeIdx - 1 - this.readOffset;
    while (idx < 0) idx += this.bufLen;
    var wet = this.buf[idx];
    this.buf[this.writeIdx] = input[i] + wet * feedback;
    this.writeIdx = (this.writeIdx + 1) % this.bufLen;
    this.readOffset++;
    if (this.readOffset >= len) this.readOffset = 0;
    output[i] = input[i] * (1 - mix) + wet * mix;
  }
  return true;
};

registerProcessor('reverse-delay', ReverseDelayProcessor);

File: plugins/delay/reverse-delay/ReverseDelay.java
package com.vocalmonitor.plugin.community;

import com.vocalmonitor.plugin.VocalMonitorNativePlugin;

// Reverse Delay (native port) — record a window of length T, play it
// back in reverse. The read head sweeps backwards through the same
// chunk the write head advances through. Pink Floyd / shoegaze swells.
// Feedback piles nested reversals into a swelling cloud.
public final class ReverseDelay implements VocalMonitorNativePlugin {
    private float[] buf;
    private int bufLen;
    private int writeIdx = 0;
    private int readOffset = 0;
    private int sampleRate = 44100;
    private float time = 600f, feedback = 0.3f, mix = 0.5f;

    @Override
    public void init(int sr) {
        this.sampleRate = sr;
        bufLen = 192000;
        buf = new float[bufLen];
        writeIdx = 0;
        readOffset = 0;
    }

    @Override public String[] parameterNames() { return new String[] { "time", "feedback", "mix" }; }
    @Override public float parameterMin(String n) {
        if ("time".equals(n)) return 100f;
        return 0f;
    }
    @Override public float parameterMax(String n) {
        switch (n) {
            case "time":     return 4000f;
            case "feedback": return 0.9f;
            default:         return 1f;
        }
    }
    @Override public float parameterDefault(String n) {
        switch (n) {
            case "time":     return 600f;
            case "feedback": return 0.3f;
            case "mix":      return 0.5f;
            default:         return 0f;
        }
    }
    @Override public String parameterLabel(String n) {
        switch (n) {
            case "time":     return "Time (ms)";
            case "feedback": return "Feedback";
            case "mix":      return "Mix";
            default:         return n;
        }
    }
    @Override public void setParameter(String n, float v) {
        switch (n) {
            case "time":     time = v; break;
            case "feedback": feedback = v; break;
            case "mix":      mix = v; break;
        }
    }

    @Override
    public void process(float[] input, float[] output) {
        int len = Math.max(64, (int) Math.floor(sampleRate * time / 1000f));
        if (len > bufLen) len = bufLen;
        final float fb = feedback;
        final float mixLocal = mix;
        final float dry = 1f - mixLocal;
        final float[] b = buf;
        final int bL = bufLen;
        int w = writeIdx;
        int ro = readOffset;
        final int n = input.length;
        for (int i = 0; i < n; i++) {
            int idx = w - 1 - ro;
            while (idx < 0) idx += bL;
            float wet = b[idx];
            b[w] = input[i] + wet * fb;
            w++; if (w >= bL) w = 0;
            ro++; if (ro >= len) ro = 0;
            output[i] = input[i] * dry + wet * mixLocal;
        }
        writeIdx = w;
        readOffset = ro;
    }
}
